### PR TITLE
[MRG] Add `--save-matches` to `fastmultigather`

### PR DESCRIPTION
**Pull request by mr-eyes**
TODO:
- [x] Implement the logic
- [x] Make it optional
- [x] Testing it
- [x] Docs

### CODE REVIEW

**Comment by mr-eyes** [comment]
Initial testing is not bad at all!

```
w/ matches=True
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  705.08s user 422.49s system 1036% cpu 1:48.81 total

w/ matches=False
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  555.32s user 5.15s system 1159% cpu 48.319 total

main branch version
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  447.15s user 66.81s system 1113% cpu 46.153 total
```

**Comment by mr-eyes** [comment]
Ready for review @ctb @bluegenes 
I couldn't test the matched hashes length because hashes can overlap across targets.

File: src/fastmultigather.rs [hunk 1]
```diff
@@ -70,12 +79,20 @@ pub fn fastmultigather(
                 let prefix = name.split(' ').next().unwrap_or_default().to_string();
                 let location = PathBuf::new(&prefix).file_name().unwrap();
                 if let Some(query_mh) = query_sig.minhash() {
+                    let mut matching_hashes = if save_matches { Vec::new() } else { Vec::new() };
```
Comment (ctb) [inline src/fastmultigather.rs:82]: this line confuses me - isn't it creating a new vec either way? What does the if/else do here?

File: src/fastmultigather.rs [hunk 2]
```diff
@@ -105,6 +122,33 @@ pub fn fastmultigather(
                             Some(gather_output),
                         )
                         .ok();
+
+                        // Save matching hashes to .sig file if save_matches is true
+                        if save_matches {
+                            let sig_filename = format!("{}.matches.sig", name);
```
Comment (ctb) [inline src/fastmultigather.rs:128]: I would suggest using .gz or zip output here. .sig format is bulky.

**Comment by ctb** [comment]
> Initial testing is not bad at all!

I'm confused as to why `--save-matches=False` is resulting in any slowdown at all!

**Comment by mr-eyes** [comment]
> > Initial testing is not bad at all!
> 
> I'm confused as to why `--save-matches=False` is resulting in any slowdown at all!

Here are the latest results after [`b546eef` (#397)](https://github.com/sourmash-bio/sourmash_plugin_branchwater/pull/397/commits/b546eef61453f7208c29ad9c31dc0ed2c3b4de8f)

```
// --save-matches=True
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  659.54s user 784.33s system 917% cpu 2:37.31 total

// --save-matches=False
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  443.37s user 3.25s system 1177% cpu 37.942 total

// main branch
sourmash scripts fastmultigather sig_paths.l hg38-entire.sig.zip -k 51  10000  453.75s user 3.63s system 1158% cpu 39.463 total
```

File: src/fastmultigather.rs (hunk 1 above)
Comment (mr-eyes) [reply src/fastmultigather.rs]: this line confused me as well - late-hour coding ;|

**Comment by ctb** [comment]
I'll take a look at this when I can, but a few quick thoughts -

* would be good to gzip the output sketches (if we're going to output a sig or sig.gz instead of a zip)
* I would like to compare to the sourmash output somehow; will think on the test

**Comment by ctb** [comment]
Testing at the command line -
```
sourmash scripts fastmultigather 47.sig.zip 47.sig.zip -s 10000 --save-matches
```
I get:
```
NC_009661.1 Shewanella baltica OS185 plasmid pS18501, complete sequence.matches.sig
NC_009661.1.gather.csv
NC_009661.1.prefetch.csv
```

so I think I'd suggest splitting off the identifier as with gather.csv and prefetch.csv.

In addition, I'd like to suggest compressing the matched sketches on writing; example code to do so appears to be over in sourmash, file `src/core/src/ffi/signature.rs`. If you don't feel like adding it please just create an issue to do so ;).

**Comment by ctb** [comment]
I've created a PR here (into your PR) with a few suggested changes.

https://github.com/sourmash-bio/sourmash_plugin_branchwater/pull/420

**Comment by ctb** [comment]
ok! a few more suggested changes & we're good to go: see https://github.com/sourmash-bio/sourmash_plugin_branchwater/pull/423

Please merge that, and then this, & then create issues as requested above ;)

**Review by ctb — APPROVED** [review]

**Comment by mr-eyes** [comment]
I will create the mentioned issues to be added later by myself or others. Many thanks!

**Pull request merged**